bsip-0044.md

```diff
@@ -0,0 +1,292 @@
+  Validate: `depositor` account has requisite `quantity` of BTS for `required_fee`
+  Validate: `recipient` account exists
+  Validate: `preimage_length` does not exceed GRAPHENE_HTLC_MAXIMUM_PREIMAGE_LENGTH
+  Validate: `preimage_hash` well formed
+  Update: BTS balance of `depositor` based on `required_fee`)
+  contract = new htlc_obj
+  Set: `contract.depositor` = `depositor`
+  Set: `contract.recipient` = `recipient`
+  Set: `contract.hash_algorithm` = `hash_algorithm`
+  Set: `contract.preimage_hash` = `preimage_hash`
+  Set: `contract.preimage_length` = `preimage_length`
+  Set: `contract.timeout_treshold` = `timeout_threshold`
+  Transfer: from `depositor` account to `contract.quantity` of `contract.symbol`
+  return results
+```
+
+### **Redeem**
+
+```
+transaction_obj htlc_redeem(fee_paying_account, id, preimage, htlc_redemption_fee)
+  Validate: transaction signed by requisite `authority` for `fee_paying_account` // any account may attempt to redeem
+  Get: get_htlc(id)
+  Validate: `fee_paying_account` account has requisite `quantity` of BTS for `htlc_redeem_fee` and `htlc_kb_fee`
+  Update: balance of `fee_paying_account` based on total fees
+  // Evaluate: timelock
+     if now() < `timeout_threshold` then return error // "timeout exceeded"
+  // Evaluate: hashlock
+     if length(preimage) != `id.preimage_length` then return error // "preimage length mismatch"
+     Calculate: `preimage_hash` = hash(preimage)
+     if `preimage_hash` != `id.preimage_hash` then return error // "invalid preimage submitted"
+     Update: balance of `id.recipient` add asset `id.symbol` of quantity `id.quantity`
+     Add: transaction to mempool 
+     Set: `id.preimage_tx_id` = `transaction_id` 
+     Cleanup: memory allocated to this htlc
+     Virtual Operation: Update account history for `depositor` to reflect redemption as by default the above operation will only appear for `redeemer`
+
+     return: results
+```
+
+### **Extend Expiry**
+
+```
+transaction_obj htlc_extend_expiry(depositor, id, timeout_threshold, htlc_extention_fee)
+  Validate: 'depositor' = get_htlc(id).depositor 
+  Validate: `timeout_threshold` < now() + GRAPHENE_HTLC_MAXIMUM_DURRATION
+  Calculate: `required_fee` = GRAPHENE_HTLC_DAILY_FEE * count((`timeout_threshold` - now()), days)
+  Validate: `depositor` account has requisite `quantity` of BTS for `required_fee`
+  Update: BTS balance of `depositor` based on `required_fee`)
+  Set: `contract.timeout_treshold` = `timeout_threshold`
+  return results
+```
+
+### **At Expiry**
+
+```
+  Get: get_htlc(id)
+  Update: balance of `depositor` add asset `id.symbol` of quantity `id.quantity`
+  Cleanup: memory allocated to this htlc
+  Virtual Operation: Update account history for `depositor` to reflect expiry without redemption.
+```
+
+## **cli_wallet APIs**
+
+### **htlc_prepare**
+
+### **htlc_redeem**
+
+### **htlc_extend_expiry**
+
+## **witness_node APIs**
+
+### **get_htlc**
```

Comment:
This call is crucial as you need to identify an HTLC in order to redeem it.
I guess, a proper naming for the call should be `get_htlcs_for_account`

Reply:
I have assumed that we would use the database ID as the identifier. This has advantages and disadvantages. The advantage is it is guaranteed to be system-wide unique after LIB. The disadvantage is that it could change before LIB, possibly misleading users. I would like to hear everyone's input on how best to implement HTLC identifiers.

Reply:
To get the object behind an idea, you can simply use `get_objects([id])`. But I may want to find (among many), the HTLC object that links to my account (as depositor or recipient).

Reply:
> The disadvantage is that it could change before then, possibly misleading user

This applies to other operations as well. Solution is to wait until LIB is past creation.
(For assets + accounts there is also assert_operation, but IMO it doesn't make sense here.)

Reply:
So I think I am hearing that the ability to retrieve all HTLCs for a particular account would be a helpful API addition. I believe that can be added without difficulty.

The harder question is if the database id is okay to use. Here is my thinking: Immediately after creation, the database id is available, but has a chance to be incorrect. I believe that the more popular HTLCs become, the greater the chance of incorrectness. Retrieving a more accurate id is a function of time. Eventually, the id is guaranteed.

Those with more experience than I would be better at guestimating when it is "mostly safe" to believe an ID. Does an ID become "mostly safe" within a reasonable period of time? I realize block times can be adjusted. I am asking for an "in practice" time period.

If we still want to use the database id, we could minimize the chance of a user looking at an incorrect contract by (as xeroc suggests) including the account name as part of the search criteria. Personally, I would rather have Bob explain to Alice that the id she shared is not valid, rather than Bob see a contract that has nothing to do with what he and Alice are negotiating.

We can walk down other schemes for ids. GUID, user assigned, hashes of a signed timestamp. But I still hope that the database id will work if it is not too inconvenient for the user.

Reply:
Is retrieving the id of the correct/safe contract an implementation detail for the client side rather than the server API side? Should we push the logic about what is valid to the client side so the end user can feel confident in the data used in their application? There may be value to querying for an contract prior to LIB.

Reply:
The real danger in publishing a claim operation is not that the ID might be wrong, the real danger is that before LIB you cannot be sure that the chain doesn't switch to a fork where the HTLC does not exists. That means you have revealed your secret, but there is nothing you can claim with it. So Alice will have to wait until Bob's HTLC is beyond LIB anyway.
Database IDs are used as a reference throughout the protocol. This is part of the reason why our transactions are so lean and our chain so fast. I see no reason to deviate from the rule in this case.

Reply:
My interpretation of @pmconrad's comment: Go with the database ID. Make sure that all developers and users of this feature are clear that a contract is not "set in stone" until the block that contains it is protected by an irreversible block.

My additional input: With the above stipulation, wallet and b2b developers that want to use this feature can add business rules and user prompting to prevent poor decisions by users / businesses.

Is that an ideal solution? I do not think so. But I doubt that an ideal solution could be worked out in a consensus system. Other chains suffer the same problem. The vast majority (all?) are worse off than Bitshares (require more time before something is considered "irreversible").

Reply:
Added method get_htlcs_for_account